### ISSUE TITLE
Feature: web/_flatten: Evaluate deferreds from Element methods asynchronously?

### ISSUE THREAD

**Issue by antler5** [twisted/twisted#12298]
> I was following along with the Twisted Templates tutorial when, instead
> of referencing an external deferred, I put the `deferLater` calls into
> the Element's `renderer` methods. In untested psudo-code, it looked
> something like this:
> 
> ```python
> sample = XMLString(
>     """
>     <div xmlns:t="http://twistedmatrix.com/ns/twisted.web.template/0.1">
>     Before waiting ... <span t:render="wait1"></span> ... after waiting.
>     Before waiting ... <span t:render="wait2"></span> ... after waiting.
>     Before waiting ... <span t:render="wait3"></span> ... after waiting.
>     </div>
>     """)
> 
> class Index(Element):
>   loader=sample
> 
>   @renderer
>   def wait1(self, request, tag):
>     return deferLater(reactor, 2, lambda: tag("first value"))
> 
>   @renderer
>   def wait2(self, request, tag):
>     return deferLater(reactor, 6, lambda: tag("second value"))
> 
>   @renderer
>   def wait3(self, request, tag):
>     return deferLater(reactor, 4, lambda: tag("third value"))
> 
> class DeferredElement(Resource):
>   def __init__(self, element):
>     Resource.__init__(self)
>     self.element = element
> 
>   def render(self, request):
>     return renderElement(request, self.element)
> 
> root = Resource()
> root.putChild(b"", DeferredElement(Index()))
> application = Site(root)
> ```
> 
> I expected each method to be ran, each deferred scheduled, and buffered
> output over the course of 6 seconds (when the latest deferred returns).
> What actually happened was that each method executed sequentially,
> "blocking" (conceptually, not in the event loop) until the prior
> method's deferred had resolved before scheduling the next.
> 
> I think it's really cool how Twisted's own templating system can
> leverage the platform, and was disappointed when it silently chained
> deferreds without explicit dependencies. I decided, primarily to learn
> more about Twisted, to make it work as I had initially expected. If you
> too would like to see Twisted Templates work this way, let me know! I'd
> be glad to come back to tidy it up in the future, but only if there's
> interest.
> 
> It's also possible that there's a bit of an X/Y problem going on, so if
> you fail to see the use case, please, feel free to elaborate on any
> evident misunderstandings or low-hanging improvements.
> 
> This commit is a first draft, having just hit the "Oh! It works!" stage.
> It's been adapted out of the existing code with minimal changes. A
> second draft would benefit from broader refactoring, tests*, and perhaps
> an extra layer of buffering to avoid creating a million tiny deferreds.
> 
> \* It was quite tricky to get output in the right order when the
> deferreds return in a different sequence, are nested, etc.

— https://github.com/antler5/twisted/commit/e949eb6f9d68d95a2cc0ea9a9ab9958d27c1c31e

Opening as an Issue rather than a PR bc, as described, it's more of a feature suggestion and feedback thing rn. Future and/or alternative PR's would be on-topic for this thread. Before I worry about documenting and eg. preventing `maybeKeepGoing` from mis-interpreting arguments, is it even something anyone would like to see fleshed out?

**Comment by adiroiban** [comment]
Hi. Thanks for the report.

I don't think that many people care about Twisted web templates these days.

I think the biggest issue here, is finding someone else who cares about Twisted web template, so that they can provide a good review.

But we should not block on this.

If you needs support for concurrent deferred and your PR has documentation and tests for this new functionality, I think that we should merge it.

I suggest to create a PR (even if just a draft) so that all the tests are executed.
As long as there are no regressions detected by the current automated tests, we should be good to go.

**Comment by glyph** [comment]
> I don't think that many people care about Twisted web templates these days.
> 
> I think the biggest issue here, is finding someone else who cares about Twisted web template, so that they can provide a good review.

I use twisted web templates and will happily provide a review.

**Comment by glyph** [comment]
> I think it's really cool how Twisted's own templating system can
> leverage the platform, and was disappointed when it silently chained
> deferreds without explicit dependencies

So, I'm not sure there's anything to be done here, although I'm certainly open to being convinced.  Renderers are all run at the point in the processing of the template where their value is needed.

You can, if you want to run things in parallel, compute attributes for your `Element` ahead of time, assign those to `Deferred`s, and return them from its various renderers.  This makes the naive renderer, one which hasn't considered in-page parallelism, work correctly, if slowly, and it's still possible to do fancy parallelism if you know you want it.

This might seem like just a bad default that you'd want to change, after all, why go slow when you could go vast, but there's also an issue of backpressure and memory consumption.  If each renderer potentially computes a large amount of data, unconditionally blasting them all out in parallel means that the rendering system has to internally catch all of their buffered output and wait for the output stream to catch up to all of them.

That said, given that there's *no* way to tell the templating system that it ought to run renderers in parallel, I can see that it is a bit limiting.  If you have a use-case for this that isn't easily satisfied by just manually coordinating the level of parallelism you want, it would definitely be a nice feature to have an option somewhere to say "every renderer in this template should be rendered as soon as it's seen by the templating system"

**Comment by antler5** [comment]
Thank you both for your perspectives! Your elaboration brings me a great deal of clarity ^u^

> As long as there are no regressions detected by the current automated tests, we should be good to go.

Not as-is lol, but I'll consider that positive feedback on the premise c:

> If you have a use-case for this that isn't easily satisfied by just manually coordinating the level of parallelism you want

Re: my own suggestion of an X/Y problem, it's quite likely that I don't have such a use-case unless I could present a clear example of it as an ergonomics thing.

> blasting them all out in parallel means that the rendering system has to internally catch all of their buffered output and wait for the output stream to catch up to all of them

> it would definitely be a nice feature to have an option somewhere

That's a really good point! And I ought have considered it, having implemented just that. I can think of small optimizations, but the back-pressure is architectural and calls for a clear and informative opt-in (or -out) mechanism for user-code (a decorator? a deferred representing the position of the output stream, to chain onto?) -- and that does sound a bit elaborate when manual management side-steps the issue. Perhaps I need a gnarly manual example (ie. a use-case >u<) to consider another approach over.

Thank you both again for your insights, I feel more confident that I've got my head around it c:
